pvlib/irradiance.py

```diff
@@ -1400,6 +1405,268 @@ def dirindex(ghi, ghi_clearsky, dni_clearsky, zenith, times, pressure=101325.,
+        Controls if the algorithm evaluates inputs with AOI >= 90 degrees.
+        If False, returns nan for AOI >= 90 degrees. Significant speed ups
+        can be achieved by setting this parameter to False.
+
+    Returns
+    -------
+    data : OrderedDict or DataFrame
+        Contains the following keys/columns:
+
+            * ``ghi``: the modeled global horizontal irradiance in W/m^2.
+            * ``dni``: the modeled direct normal irradiance in W/m^2.
+            * ``dhi``: the modeled diffuse horizontal irradiance in
+              W/m^2.
+
+    References
+    ----------
+    .. [1] B. Marion, A model for deriving the direct normal and
+           diffuse horizontal irradiance from the global tilted
+           irradiance, Solar Energy 122, 1037-1046.
+           http://dx.doi.org/10.1016/j.solener.2015.10.024
+    """
+
+    aoi_lt_90 = aoi < 90
+
+    # for AOI less than 90 degrees
+
+    ghi, dni, dhi, kt_prime = _gti_dirint_lt_90(
+        poa_global, aoi, aoi_lt_90, solar_zenith, solar_azimuth,
+        surface_tilt, surface_azimuth, times, pressure=pressure,
+        use_delta_kt_prime=use_delta_kt_prime, temp_dew=temp_dew,
+        albedo=albedo, model=model, model_perez=model_perez,
+        max_iterations=max_iterations)
+
+    # for AOI greater than or equal to 90 degrees
+
+    if calculate_gt_90:
+        ghi_gte_90, dni_gte_90, dhi_gte_90 = _gti_dirint_gte_90(
+            poa_global, aoi, solar_zenith, solar_azimuth,
+            surface_tilt, times, kt_prime,
+            pressure=pressure, temp_dew=temp_dew, albedo=albedo)
+    else:
+        ghi_gte_90, dni_gte_90, dhi_gte_90 = np.nan, np.nan, np.nan
+
+    # put the AOI < 90 and AOI >= 90 conditions together
+
+    output = OrderedDict()
+    output['ghi'] = ghi.where(aoi_lt_90, ghi_gte_90)
+    output['dni'] = dni.where(aoi_lt_90, dni_gte_90)
+    output['dhi'] = dhi.where(aoi_lt_90, dhi_gte_90)
+
+    output = pd.DataFrame(output, index=times)
+
+    return output
+
+
+def _gti_dirint_lt_90(poa_global, aoi, aoi_lt_90, solar_zenith, solar_azimuth,
+                      surface_tilt, surface_azimuth,
+                      times, pressure=101325.,
+                      use_delta_kt_prime=True, temp_dew=None, albedo=.25,
+                      model='perez', model_perez='allsitescomposite1990',
+                      max_iterations=30):
+
+    I0 = extraradiation(times, 1370, 'spencer')
+    # cos_zenith = np.maximum(0.065, tools.cosd(solar_zenith))
+    cos_zenith = tools.cosd(solar_zenith)
+    I0h = I0 * cos_zenith
+
+    # these coeffs and diff variables and the loop below
+    # implement figure 1 of Marion 2015
+
+    # make coeffs that is at least 30 elements long
+    # for loop below will limit iterations if necessary
+    coeffs = np.empty(max(30, max_iterations))
+    coeffs[0:3] = 1
+    coeffs[3:10] = 0.5
+    coeffs[10:20] = 0.25
+    coeffs[20:] = 0.125
+
+    # initialize diff
+    diff = pd.Series(9999, index=times)
```

Comment:
Would it be a good idea to convert any remaining 999 to nan at the end?

Reply:
to convert the ghi, dni, dhi to nan where diff remains 9999? alternatively, we could do this for any point that does not converge. This seems to easily occur near aoi = 90.